File: rust/src/picnic.rs


const ARE_FRIENDS: [[bool; 10]; 10] = [
    [true, true, true, true, true, true, true, true, true, true], // 0
    [true, true, true, true, true, true, true, true, true, true], // 1
    [true, true, true, true, true, true, true, true, true, true], // 2
    [true, true, true, true, true, true, true, true, true, true], // 3
    [true, true, true, true, true, true, true, true, true, true], // 4
    [true, true, true, true, true, true, true, true, true, true], // 5
    [true, true, true, true, true, true, true, true, true, true], // 6
    [true, true, true, true, true, true, true, true, true, true], // 7
    [true, true, true, true, true, true, true, true, true, true], // 8
    [true, true, true, true, true, true, true, true, true, true], // 9
];


pub fn count_pairings(taken: Vec<bool>) -> usize {
    let first_free = taken.iter().position(|&e| e == false);
    match first_free {
        None             => 1,
        Some(first_free) => (first_free+1..taken.len())
            .map(|pair_with| 
                match (taken[pair_with], ARE_FRIENDS[first_free][pair_with]) { 
                    (has_pair, is_friend ) if !has_pair && is_friend => { 
                        let mut new_taken = taken.clone();
                        new_taken[first_free] = true;
                        new_taken[pair_with] = true;
                        count_pairings(new_taken)
                    }, // match
                    _ => 0 
                }) // map 
            .sum(),
    } // match 
}

File: haskell/Picnic.hs

module Picnic where


    import Control.Monad
    import Data.List
    import Debug.Trace


    randomFriends :: [[Bool]]
    randomFriends = [[ True, False, False,  True,  True,  True,  True,  True, False,  True] -- 0
                    ,[ True, False,  True, False,  True, False,  True, False, False, False] -- 1
                    ,[False, False,  True, False, False, False, False, False, False, False] -- 2
                    ,[False, False,  True, False, False,  True, False, False, False,  True] -- 3
                    ,[False,  True,  True, False,  True, False,  True, False,  True,  True] -- 4
                    ,[ True,  True,  True, False,  True,  True, False,  True, False, False] -- 5
                    ,[False, False,  True,  True, False, False,  True,  True,  True,  True] -- 6
                    ,[False, False,  True,  True,  True, False,  True,  True, False, False] -- 7
                    ,[ True, False, False, False, False, False, False, False, False, False] -- 8
                    ,[ True, False,  True,  True,  True, False,  True, False,  True,  True] -- 9
                    ]


    allFriends :: [[Bool]]
    allFriends = [[ True,  True,  True,  True,  True,  True,  True,  True,  True,  True] -- 0
                 ,[ True,  True,  True,  True,  True,  True,  True,  True,  True,  True] -- 1
                 ,[ True,  True,  True,  True,  True,  True,  True,  True,  True,  True] -- 2
                 ,[ True,  True,  True,  True,  True,  True,  True,  True,  True,  True] -- 3
                 ,[ True,  True,  True,  True,  True,  True,  True,  True,  True,  True] -- 4
                 ,[ True,  True,  True,  True,  True,  True,  True,  True,  True,  True] -- 5
                 ,[ True,  True,  True,  True,  True,  True,  True,  True,  True,  True] -- 6
                 ,[ True,  True,  True,  True,  True,  True,  True,  True,  True,  True] -- 7
                 ,[ True,  True,  True,  True,  True,  True,  True,  True,  True,  True] -- 8
                 ,[ True,  True,  True,  True,  True,  True,  True,  True,  True,  True] -- 9
                 ]


    areFriends = allFriends

    countPairings :: [Bool] -> Maybe Int
    countPairings taken = do
        let firstFree = elemIndex False taken
        case firstFree of
          Nothing   -> return 1
          Just firstFree' ->
              fmap sum $ forM [firstFree'+1..n] $ \ pairWith -> do
                  case (not $ taken!!pairWith) && (areFriends!!firstFree'!!pairWith) of
                    True  -> countPairings [if i == firstFree' || i == pairWith 
                                              then True 
                                              else t 
                                                | (i, t) <- zip [0..n] taken]
                    False -> return 0  
        where 
            n = (length taken) - 1


